backend/src/main/java/pt/ist/socialsoftware/mono2micro/utils/Metrics.java

```diff
@@ -25,38 +25,106 @@ public void calculateMetrics() {
 			graphComplexity += controller.getComplexity();
-			calculateClusterDependencies(controller);
 		}
 
-		graphComplexity /= graph.getControllers().size();
+		graphComplexity /= graphControllers.size();
 		graphComplexity = BigDecimal.valueOf(graphComplexity).setScale(2, RoundingMode.HALF_UP).floatValue();
+
 		this.graph.setComplexity(graphComplexity);
 
-		for (Cluster cluster : graph.getClusters()) {
-			calculateClusterComplexity(cluster);
+		List<Cluster> graphClusters = graph.getClusters();
+
+		for (Cluster cluster : graphClusters) {
+			calculateClusterComplexityAndCohesion(cluster);
 
-			calculateClusterCohesion(cluster);
 			graphCohesion += cluster.getCohesion();
 
 			calculateClusterCoupling(cluster);
 			graphCoupling += cluster.getCoupling();
 		}
 
-		graphCohesion /= graph.getClusters().size();
+		int graphClustersAmount = graphClusters.size();
+
+		graphCohesion /= graphClustersAmount;
 		graphCohesion = BigDecimal.valueOf(graphCohesion).setScale(2, RoundingMode.HALF_UP).floatValue();
 		this.graph.setCohesion(graphCohesion);
-		graphCoupling /= graph.getClusters().size();
+
+		graphCoupling /= graphClustersAmount;
 		graphCoupling = BigDecimal.valueOf(graphCoupling).setScale(2, RoundingMode.HALF_UP).floatValue();
 		this.graph.setCoupling(graphCoupling);
     }
 
+    private void calculateControllerComplexityAndClusterDependencies(Controller controller) {
+		Set<Controller.LocalTransaction> allLocalTransactions = controller.getAllLocalTransactions();
+
+		if (this.controllerClusters.get(controller.getName()).size() == 1) {
+			controller.setComplexity(0);
+
+			// I'm so sorry but... is there a better way?
+			for (Controller.LocalTransaction lt : allLocalTransactions) {
+				// ClusterDependencies
+				Cluster fromCluster = graph.getCluster(lt.getClusterName());
+
+				if (fromCluster != null) { // not root node
+					List<Controller.LocalTransaction> nextLocalTransactions = controller.getNextLocalTransactions(lt);
+
+					for (Controller.LocalTransaction nextLt : nextLocalTransactions) {
+						String toEntity = nextLt.getClusterAccesses().get(0).getEntity();
+						fromCluster.addCouplingDependency(nextLt.getClusterName(), toEntity);
+					}
+				}
+			}
+
+		} else {
+
+			Map<String, List<String>> cache = new HashMap<>(); // < entity + mode, List<controllerName>> controllersThatTouchSameEntities for a given mode
+			float controllerComplexity = 0;
+
+			for (Controller.LocalTransaction lt : allLocalTransactions) {
+				// ClusterDependencies
+				Cluster fromCluster = graph.getCluster(lt.getClusterName());
+
+				if (fromCluster != null) { // not root node
+					List<Controller.LocalTransaction> nextLocalTransactions = controller.getNextLocalTransactions(lt);
+
+					for (Controller.LocalTransaction nextLt : nextLocalTransactions) {
+						String toEntity = nextLt.getClusterAccesses().get(0).getEntity();
+						fromCluster.addCouplingDependency(nextLt.getClusterName(), toEntity);
+					}
+				}
```

Comment:
O caso que falei que foi mal refatorizado.. Não é suposto ser um for em todas a localTransactions seguintes, é suposto ir buscar apenas a seguinte e adicionar a dependencia de uma para a outra